web/packages/teleterm/src/ui/DocumentAuthorizeWebSession/DocumentAuthorizeWebSession.tsx

```diff
@@ -0,0 +1,188 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { Text, Alert, ButtonPrimary, H1, ButtonText } from 'design';
+import Flex from 'design/Flex';
+import { useAsync, Attempt } from 'shared/hooks/useAsync';
+import { processRedirectUri } from 'shared/redirects';
+import { Cluster } from 'gen-proto-ts/teleport/lib/teleterm/v1/cluster_pb';
+import { DeviceConfirmationToken } from 'gen-proto-ts/teleport/devicetrust/v1/device_confirmation_token_pb';
+
+import Document from 'teleterm/ui/Document';
+import { routing } from 'teleterm/ui/uri';
+import { retryWithRelogin } from 'teleterm/ui/utils';
+import { useAppContext } from 'teleterm/ui/appContextProvider';
+import * as types from 'teleterm/ui/services/workspacesService';
+import { WebSessionRequest } from 'teleterm/ui/services/workspacesService';
+import { useWorkspaceContext } from 'teleterm/ui/Documents';
+
+export function DocumentAuthorizeWebSession(props: {
+  doc: types.DocumentAuthorizeWebSession;
+  visible: boolean;
+}) {
+  const ctx = useAppContext();
+  const { documentsService } = useWorkspaceContext();
+  const rootCluster = ctx.clustersService.findCluster(props.doc.rootClusterUri);
+  const [attempt, authorize] = useAsync(async () => {
+    const {
+      response: { confirmationToken },
+    } = await retryWithRelogin(ctx, props.doc.rootClusterUri, () =>
+      ctx.clustersService.authenticateWebDevice(
+        props.doc.rootClusterUri,
+        props.doc.webSessionRequest
+      )
+    );
+    const url = buildAuthorizedSessionUrl(
+      rootCluster,
+      props.doc.webSessionRequest,
+      confirmationToken
+    );
+    // This endpoint verifies the token and "upgrades" the web session and redirects to "/web".
+    window.open(url);
+  });
+  const clusterName = routing.parseClusterName(props.doc.rootClusterUri);
+  const canAuthorize = rootCluster.loggedInUser?.isDeviceTrusted;
+
+  async function authorizeAndCloseDocument() {
+    const [, error] = await authorize();
+    if (!error) {
+      closeAndNotify();
+    }
+  }
+
+  function openUnauthorizedAndCloseDocument() {
+    const url = buildUnauthorizedSessionUrl(
+      rootCluster,
+      props.doc.webSessionRequest
+    );
+    window.open(url);
+    closeAndNotify();
+  }
+
+  function closeAndNotify() {
+    documentsService.close(props.doc.uri);
+    ctx.notificationsService.notifyInfo(
+      'Web session has been opened in the browser'
+    );
+  }
+
+  return (
+    <Document visible={props.visible}>
+      <Flex
+        flexDirection="column"
+        maxWidth="680px"
+        width="100%"
+        mx="auto"
+        mt="4"
+        px="5"
+      >
+        <H1 mb="4">Authorize Web Session</H1>
+        <Flex flexDirection="column" gap={3}>
+          {!canAuthorize && (
```

Comment:
It wasn't necessary, but I thought it's good to gracefully handle this case.

Reply:
I meant that checking `canAuthorize` is not needed to fix the linked issue, but since we have a document and a place to show it, we can slightly improve UX.